src/compiler/compiler.rs

```diff
@@ -875,16 +888,36 @@ where
     trace!("detect_compiler: {}", executable.display());
 
     // First, see if this looks like rustc.
-    let filename = match executable.file_stem() {
-        None => bail!("could not determine compiler kind"),
-        Some(f) => f,
+
+    let rustc_executable = if is_rustc_like(executable) {
+        Some(executable.to_path_buf())
+    } else if env.iter().any(|(k, _)| k == OsStr::new("CARGO")) {
+        // If not, detect the scenario where cargo is configured to wrap rustc with something other than sccache.
+        // This happens when sccache is used as a RUSTC_WRAPPER and another tool is used as a
+        // RUSTC_WORKSPACE_WRAPPER. In that case rustc will be the first argument rather than the command.
+        //
+        // The check for the CARGO env acts as a guardrail against false positives.
+        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
```

Comment:
Not entirely sure we want to bind ourselves too much to cargo _exclusively_. Other build systems might invoke `rustc`
more directly and hence the caching calls would not work as anticipated. Since this is very much an edge case,
an issue would be sufficient I think describing the current "limitation".

Reply:
I've documented this at https://github.com/mozilla/sccache/issues/1314